### PR TITLE
[docs] Custom components page

### LINKED ISSUE #1767: Add page on setting up your torchtune project and launching custom recipes
There's no discussion on the general workflow of setting up your own project that uses torchtune and launching custom configs/recipes and custom components. Most users end up just cloning the repo and making modification directly, which is not the workflow we want to encourage. We should have a page under Basics or Getting started that discusses this.

### PR DESCRIPTION
**Pull request by RdoubleA**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

Closes #1767.

#### Changelog
What are the changes made in this PR?
*

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: docs/source/basics/custom_components.rst [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        if packed:
+            return PackedDataset(ds, max_seq_len=tokenizer.max_seq_len, split_across_pack=False)
+        else:
+            return ds
+
+.. note::
+
+    If you are using a default torchtune recipe with a custom dataset, you must define the first
+    positional argument to be the tokenizer or model transform. These are automatically passed into
+    dataset instantiation and are defined separately in the config, not under the dataset field.
```
Comment (SalmanMohammadi) [inline docs/source/basics/custom_components.rst:144]:
```suggestion
    dataset during instantiation and are defined separately in the config, not under the dataset field.
```

File: docs/source/basics/custom_components.rst [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+you are launching with ``tune run``. It is best to define the path relative to your project root directory
+and launch from there.
+
+.. code-block:: yaml
+
+    # In YAML file config/custom_finetune.yaml
+    model:
+      _component_: models.custom_decoder.custom_model
+      num_layers: 32
+
```
Comment (SalmanMohammadi) [inline docs/source/basics/custom_components.rst:156]:
```suggestion
      num_layers: 32
      classification_head: False

```

Comment (SalmanMohammadi) [reply docs/source/basics/custom_components.rst:156]: just to be super obvious what we're doing here

File: docs/source/basics/custom_components.rst [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        """
+        Python subset of nampdn-ai/tiny-codes. Instruct and code response pairs.
+        """
+        ds = SFTDataset(
+            model_transform=tokenizer,
+            source="nampdn-ai/tiny-codes",
+            message_transform=InputOutputToMessages(
+                column_map={"input": "prompt", "output": "response"},
+            ),
+            filter_fn=lambda x: x["language"] == "python",
```
Comment (SalmanMohammadi) [inline docs/source/basics/custom_components.rst:132]: this is a really slick example

File: docs/source/basics/custom_components.rst [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+.. code-block:: yaml
+
+    # In YAML file config/custom_finetune.yaml
+    model:
+      _component_: models.custom_decoder.custom_model
+      num_layers: 32
+
+    dataset:
+      _component_: datasets.custom_dataset.tiny_codes
+      packed: True
```
Comment (SalmanMohammadi) [inline docs/source/basics/custom_components.rst:159]:
```suggestion
      # we don't need to define a tokenizer here as it's automatically passed in
      packed: True
```

File: docs/source/basics/custom_components.rst (hunk 3 above)
Comment (RdoubleA) [reply docs/source/basics/custom_components.rst:132]: I actually used this too so it's ~~authentic~~

File: docs/source/index.rst [hunk 5]
```diff
@@ -131,6 +131,7 @@ torchtune tutorials.
    basics/message_transforms
    basics/tokenizers
    basics/prompt_templates
+   basics/custom_components
    basics/packing
```
Comment (SalmanMohammadi) [inline docs/source/index.rst:135]: GIGA NIT
```suggestion
   basics/packing
   basics/custom_components
```

File: docs/source/basics/custom_components.rst (hunk 2 above)
Comment (RdoubleA) [reply docs/source/basics/custom_components.rst:156]: wanted to subtly show that this is an optional param so didn't need to specify in config. but I'm ok either way

File: docs/source/index.rst (hunk 5 above)
Comment (RdoubleA) [reply docs/source/index.rst:135]: lordy lord

**Comment by SalmanMohammadi** [comment]
If you can be bothered I think this belongs in the README, somewhere like https://github.com/pytorch/torchtune?tab=readme-ov-file#custom-datasets

File: docs/source/basics/custom_components.rst (hunk 2 above)
Comment (SalmanMohammadi) [reply docs/source/basics/custom_components.rst:156]:
```
# this is an optional param, so you can also omit this from the config
```
mayhaps

File: docs/source/basics/custom_components.rst [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+=============================
+
+torchtune lets you launch fine-tuning jobs directly from the command-line using both library and custom datasets,
+models, recipes, and configs. This is done with the ``tune run`` command (see :ref:`cli_label`), which can also be
+used from your project folder.
+
+Setting up your torchtune project
+---------------------------------
+First, ensure that you have torchtune installed, see :ref:`install_label`. This will also install the ``tune`` command
+to your environment.
```
Comment (SalmanMohammadi) [inline docs/source/basics/custom_components.rst:14]:
```suggestion
First, ensure that you have torchtune installed, see :ref:`install_label`. This will also enable the ``tune`` command
in your environment.
```

File: docs/source/basics/custom_components.rst [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+You can launch with ``tune run`` from any directory. Let's create a new project directory and ensure we can launch a built-in
```
Comment (SalmanMohammadi) [inline docs/source/basics/custom_components.rst:16]:
```suggestion
You can now use ``tune run`` from any directory. Let's create a new project directory and ensure we can launch a built-in
```

File: docs/source/basics/custom_components.rst [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+Launching custom configs
+------------------------
+Often, you'll want to start with one of our default configs for a particular model and adjust a few training hyperparameters.
+You can use the ``tune cp`` command to create a copy of a default config in your project directory so you can make modifications.
+
+.. code-block:: bash
+
+    # Show all the default model configs for each recipe
+    tune ls
```
Comment (SalmanMohammadi) [inline docs/source/basics/custom_components.rst:36]: tune ls output be like 
image

File: docs/source/basics/custom_components.rst [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    def main(cfg: DictConfig):
+        # Add all your recipe logic here, access config fields as attributes
+
+    if __name__ == "__main__":
+        # Config will be parsed from CLI, don't need to pass in here
+        main()
+
+Launching with custom components
+--------------------------------
+torchtune supports full experimentation with custom models, datasets, optimizers, or any fine-tuning component. You can define
```
Comment (SalmanMohammadi) [inline docs/source/basics/custom_components.rst:89]:
```suggestion
torchtune enables full experimentation with custom models, datasets, optimizers, or any other fine-tuning component. You can define
```

File: docs/source/basics/custom_components.rst (hunk 8 above)
Comment (RdoubleA) [reply docs/source/basics/custom_components.rst:36]: more like
image

File: docs/source/basics/custom_components.rst [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+    class CustomTransformerDecoder(nn.Module):
+        # A custom architecture not present in torchtune
+
+    # Builder function for the custom model
+    def custom_model(num_layers: int, classification_head: bool = False):
+        # Any setup for defining the class
+        ...
+        # Return the module you want to train
+        return CustomTransformerDecoder(...)
+
```
Comment (SalmanMohammadi) [inline docs/source/basics/custom_components.rst:110]:
```suggestion

This allows us to expose our custom model in a config friendly manner - rather than having to re-define every argument needed to construct our custom model, we only expose the arguments which we care about modifying. This is how we implement our models in torchtune - see :func:`torchtune.models.llama3_2_vision.llama3_2_vision_11b` as an example!

```
just a suggestion to help be super obvious why we reccomend doing it this way, I'm not sure if everyone would clock this design decision immediately

File: docs/source/basics/custom_components.rst [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+models, recipes, and configs. This is done with the ``tune run`` command (see :ref:`cli_label`), which can also be
+used from your project folder.
+
+Setting up your torchtune project
+---------------------------------
+First, ensure that you have torchtune installed, see :ref:`install_label`. This will also install the ``tune`` command
+to your environment.
+
+You can launch with ``tune run`` from any directory. Let's create a new project directory and ensure we can launch a built-in
+library recipe with a library config from that folder.
```
Comment (SalmanMohammadi) [inline docs/source/basics/custom_components.rst:17]:
```suggestion
First, ensure that you have torchtune installed - see :ref:`install_label`. This will install the ``tune`` command
in your environment, so you can launch ``tune run`` from any directory. Let's create a new project directory and ensure we can launch a built-in
library recipe with a library config from that folder.
```
idk I'm being opinionated

**Review by SalmanMohammadi — APPROVED** [review]
